sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScalerProvider.cs

```diff
@@ -0,0 +1,148 @@
+
+        public IEnumerable<IScaleMonitor> GetScaleMonitors()
+        {
+            foreach (var triggerMetada in _triggerMetadata.Where(x => x.Type == "serviceBusTrigger"))
+            {
+                yield return GetMonitor(triggerMetada);
+            }
+        }
+
+        public IEnumerable<ITargetScaler> GetTargetScalers()
+        {
+            foreach (var triggerMetada in _triggerMetadata.Where(x => x.Type == "serviceBusTrigger"))
+            {
+                yield return GetTargetScaler(triggerMetada);
+            }
+        }
+
+        private IScaleMonitor GetMonitor(TriggerMetadata triggerMetadata)
+        {
+            ServiceBusMetadata serviceBusMetadata = JsonConvert.DeserializeObject<ServiceBusMetadata>(triggerMetadata.Metadata.ToString());
+            AzureComponentFactory azureComponentFactory = GetAzureComponentFactory(triggerMetadata);
+
+            (ServiceBusAdministrationClient adminClient, ServiceBusReceiver receiver) = CreateParameters(_configuration, serviceBusMetadata, azureComponentFactory);
+
+            return new ServiceBusScaleMonitor(
+                triggerMetadata.FunctionName,
+                receiver.EntityPath,
+                !string.IsNullOrEmpty(serviceBusMetadata.QueueName) ? ServiceBusEntityType.Queue : ServiceBusEntityType.Topic,
+                new Lazy<ServiceBusReceiver>(() => receiver),
+                new Lazy<ServiceBusAdministrationClient>(() => adminClient),
+                _loggerFactory);
+        }
+
+        private ITargetScaler GetTargetScaler(TriggerMetadata triggerMetadata)
+        {
+            ServiceBusMetadata serviceBusMetadata = JsonConvert.DeserializeObject<ServiceBusMetadata>(triggerMetadata.Metadata.ToString());
+            AzureComponentFactory azureComponentFactory = GetAzureComponentFactory(triggerMetadata);
+
+            (ServiceBusAdministrationClient adminClient, ServiceBusReceiver receiver) = CreateParameters(_configuration, serviceBusMetadata, azureComponentFactory);
+
+            return new ServiceBusTargetScaler(
+                triggerMetadata.FunctionName,
+                receiver.EntityPath,
+                receiver.EntityPath.Contains("//") ? ServiceBusEntityType.Topic : ServiceBusEntityType.Queue,
+                new Lazy<ServiceBusReceiver>(() => receiver),
+                new Lazy<ServiceBusAdministrationClient>(() => adminClient),
+                _options.Value,
+                serviceBusMetadata.IsSessionsEnabled,
+                !string.Equals(serviceBusMetadata.Cardinality, "many"),
+                _loggerFactory);
+        }
+
+        private (ServiceBusAdministrationClient AdminClient, ServiceBusReceiver Receiver) CreateParameters(IConfiguration configuration, ServiceBusMetadata serviceBusMetadata, AzureComponentFactory azureComponentFactory)
+        {
+            ServiceBusClientFactory factory = new ServiceBusClientFactory(configuration, azureComponentFactory, _messagingProvider, _logForwarder, _options);
```

Comment:
Yes, this shows the drawback of flowing the AzureComponentFactory override as an object property. We basically have to ignore the ServiceBusClientFactory we've registered via DI for normal flows, and recreate a new instance explicitly not via DI. That could get messy since other components are using the default factory - if those components get used for anything they're using the wrong factory.

An alternative we could consider would be to embrace the fact that we're saying the component factory to use may now vary based on context in some cases - i.e. not a single instance for all usages in a host. To make that explicit we could introduce an interface like the following:

```csharp
IAzureComponentFactoryProvider
  AzureComponentFactory GetFactory(AzureComponentFactoryProviderContext context)

AzureComponentFactoryProvoderContext
  string FunctionId
```

- this would live in  the WebJobs Host.Storage package where AddAzureStorageCoreServices lives
- we'd register a default implementation of this that returns the default AzureComponentFactory in all cases regardless of context
- ScaleHosts would provide a custom implementation that returns function specific component factory instances based on context.
- in ServiceBusExtension, ServiceBusClientFactory gets updated to take an IAzureComponentFactoryProvider on construction. I.e. all cases where we're taking AzureComponentFactory via DI we replace with this interface
- There would then be some cascading changes because callers creating a client will need to flow function information down

This has the benefits of avoiding the issues mentioned above, and also means we don't need the Properties object hack on functions metadata for flowing this.